sdk/core/core/src/az_span.c

```diff
@@ -36,15 +39,44 @@ AZ_NODISCARD az_span az_span_from_str(char* str)
 
 AZ_NODISCARD az_span az_span_slice(az_span span, int32_t low_index, int32_t high_index)
 {
-  AZ_PRECONDITION_RANGE(-1, high_index, az_span_capacity(span));
-  AZ_PRECONDITION(high_index == -1 || (high_index >= 0 && low_index <= high_index));
+  // The following set of preconditions validate that:
+  //    -1 <= high_index <= span.capacity
+  // And
+  //    If high_index == -1:
+  //        0 <= low_index <= span.capacity
+  //    Otherwise
+  //        0 <= low_index <= high_index
+  AZ_PRECONDITION_INT32_RANGE(-1, high_index, az_span_capacity(span));
+
+  // Do not reorder the OR conditions, since we are relying on short-circuiting
+  AZ_PRECONDITION(
+      (high_index == -1 && ((uint32_t)low_index <= (uint32_t)az_span_capacity(span)))
+      || ((uint32_t)low_index <= (uint32_t)high_index));
 
   int32_t const capacity = az_span_capacity(span);
 
   high_index = high_index == -1 ? capacity : high_index;
   return az_span_init(az_span_ptr(span) + low_index, high_index - low_index, capacity - low_index);
 }
 
+AZ_NODISCARD az_span az_span_slice_start(az_span span, int32_t low_index)
```

Comment:
@JeffreyRichter - what are your thoughts on explicitly adding this slice overload that doesn't take a `high_index` and remove the concept of having `high_index = -1` to mean `span.capacity`?

Reply:
I prefer to leave as is.
When you start introducing multiple functions that logically do the same thing, users must know them all, understand them all, and then choose the correct one. This puts cognitive overload on the customer. And, if they start using one and then modify their code to need something else, then the modify isn't just a small change (like modifying a an argument; now they have to change the function name AND reconsider all the arguments. Right now, we have the concept of slicing a span and there is 1 function to slice a span. And to slice a span, you NEED a start and an end. For convenience we allow an end of -1 to mean capacity but this is easily changeable in customer code if they need to change it.

FYI: We used to have function to slice form the beginning and slice from the end but we got rid of them.